[CORE-8161] `storage`: add `min.cleanable.dirty.ratio`, schedule compaction by `dirty_ratio`

WillemKauf opened the pull request:
Based on PR https://github.com/redpanda-data/redpanda/pull/24649.

Instead of unconditionally compacting all logs during a round of housekeeping, users may now optionally schedule log compaction in the `log_manager` using the cluster/topic property `min_cleanable_dirty_ratio`/`min.cleanable.dirty.ratio`.

As mentioned in the above PR,

> The dirty ratio of a log is defined as the ratio between the number of bytes in "dirty" segments and the total number of bytes in closed segments.
> Dirty segments are closed segments which have not yet been cleanly compacted- i.e, duplicates for keys in this segment could be found in the prefix of the log up to this segment.

By setting the `min.cleanable.dirty.ratio` on a per topic basis, users can avoid unnecessary read/write amplification during compaction as the log grows in size.

A housekeeping scan will still be performed every `log_compaction_interval_ms`, and the log's `dirty_ratio` will be tested against `min.cleanable.dirty.ratio` in determining it's eligibility for compaction. Additionally, logs are now compacted in descending order according to their dirty ratio, offering a better "bang for buck" heuristic for compaction scheduling.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Allow for optional scheduling of compaction via `min_cleanable_dirty_ratio`

WillemKauf committed:
`cluster`: add `min_cleanable_dirty_ratio` to `topic_properties`

WillemKauf committed:
`cluster`: add `min_cleanable_dirty_ratio` to `ntp_config`

WillemKauf committed:
`config`: add `min_cleanable_dirty_ratio` to `configuration`

WillemKauf committed:
`kafka`: add `min.cleanable.dirty.ratio` to topic config handlers

WillemKauf committed:
`storage`: make `dirty_ratio` a pure `virtual` function in `log`

WillemKauf committed:
`storage`: schedule compaction by `dirty_ratio`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61464-->
<!--8cba215f-eb7a-4715-87b0-8537d2f5dfce-->
<!--cf08abdd-4ccb-4c56-8f7d-37c460b55529-->
<!--title-->
## Retry command for [Build#61464](https://buildkite.com/redpanda/redpanda/builds/61464)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/tests/partition_force_reconfiguration_test.py::NodeWiseRecoveryTest.test_node_wise_recovery@{"dead_node_count":2}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--135664cf-91a6-41d0-b8c3-8c3d340a726e-->
<!--TEST RESULTS-->
<!--11889f3e-d284-49da-9633-7c0b6c73fcbf-->
<!--TEST RESULTS-->
<!--f6bdcae8-9461-4a65-8dfa-0c087b038eac-->
<!--TEST RESULTS-->
<!--aaea2f35-0d76-4edb-9d70-762a83730431-->
<!--TEST RESULTS-->
<!--0e1cfe14-e3bc-4f1d-87ce-85f027a32e50-->
<!--TEST RESULTS-->
<!--a6fce2c9-df5b-47fd-a537-25ac19adde84-->
<!--TEST RESULTS-->
<!--f8321237-4f65-4190-a59c-525f3a7904b5-->
<!--TEST RESULTS-->
<!--350015a4-716b-4e53-b2be-e2e29f76490b-->
<!--TEST RESULTS-->
<!--41466236-7a8f-4c87-a66f-ee4c7dbe5cab-->
<!--TEST RESULTS-->
<!--36799ea6-badd-4df6-b86c-5356f2b67096-->
<!--TEST RESULTS-->
<!--1c40390b-5851-40fd-ace1-af973b05f699-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61464">build#61464</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61464#0194be30-4fe3-4201-a506-11eca43a36e9 | FLAKY | 1/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61464#0194be30-4fe3-4201-a506-11eca43a36e9 | FAIL | 0/2 |
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61464#0194be87-f8fb-4fbd-9207-0641084924fe | FLAKY | 1/2 |
| rptest.tests.describe_topics_test.DescribeTopicsTest.test_describe_topics_with_documentation_and_types | ducktape | https://buildkite.com/redpanda/redpanda/builds/61464#0194be87-f8f9-4197-b75d-9e2b0e669087 | FAIL | 0/20 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndThrottlingTest.test_throttling.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61464#0194be87-f8fb-4fbd-9207-0641084924fe | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.NodeWiseRecoveryTest.test_node_wise_recovery.dead_node_count=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61464#0194be87-f8fb-4fbd-9207-0641084924fe | FAIL | 0/20 |
| rptest.tests.retention_policy_test.RetentionPolicyTest.test_changing_topic_retention_with_restart.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61464#0194be87-f8f9-4197-b75d-9e2b0e669087 | FLAKY | 1/3 |
| test_compat_rpunit.test_compat_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61464#0194be2d-b364-4273-8107-604c1a24978e | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61480">build#61480</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61480#0194c30c-66c7-425f-882b-a62f163db808 | FLAKY | 1/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61480#0194c30c-66c8-4db9-a2da-7dc5f0262638 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndThrottlingTest.test_throttling.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61480#0194c368-0af7-4d62-86e0-9ef129bfeb88 | FLAKY | 1/2 |
| rptest.tests.offset_for_leader_epoch_archival_test.OffsetForLeaderEpochArchivalTest.test_querying_archive | ducktape | https://buildkite.com/redpanda/redpanda/builds/61480#0194c368-0af6-4ff1-92d6-ad77df6dc1f7 | FLAKY | 1/5 |
| rptest.tests.partition_force_reconfiguration_test.NodeWiseRecoveryTest.test_node_wise_recovery.dead_node_count=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61480#0194c368-0af7-4d62-86e0-9ef129bfeb88 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61769">build#61769</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61769#0194ed98-e02d-49fb-815b-8a92ef12fe4a | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61769#0194ed9d-c07e-41c7-ad37-873a07105423 | FLAKY | 1/2 |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_consumer_group_mirroring.source_type=kafka | ducktape | https://buildkite.com/redpanda/redpanda/builds/61769#0194ed98-e02c-45df-bba3-2bf6f621197c | FAIL | 0/20 |
| rptest.tests.partition_movement_test.PartitionMovementTest.test_availability_when_one_node_down | ducktape | https://buildkite.com/redpanda/redpanda/builds/61769#0194ed98-e02c-45df-bba3-2bf6f621197c | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61775">build#61775</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61775#0194effa-4c71-49e2-a5c3-06cce4574f93 | FLAKY | 1/2 |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_consumer_group_mirroring.source_type=kafka | ducktape | https://buildkite.com/redpanda/redpanda/builds/61775#0194eff5-6746-4bf1-aa34-e78dd6b67522 | FAIL | 0/20 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61775#0194effa-4c73-4464-9bfb-d54df885c3c0 | FLAKY | 1/2 |
| test_compat_rpunit.test_compat_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61775#0194efb0-cf5d-48a4-9e93-707da0399798 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61791">build#61791</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61791#0194f362-3879-4838-b35b-39ecd5051d85 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61791#0194f362-387a-4e5d-9f71-62729ec6a9da | FLAKY | 1/4 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61791#0194f362-387b-41cd-954d-e30904bab43d | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61791#0194f362-387b-41cd-954d-e30904bab43d | FLAKY | 1/4 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61791#0194f362-3879-4838-b35b-39ecd5051d85 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61800">build#61800</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f54b-039e-4ea0-858f-4f25d07def12 | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f539-0f33-48da-a1a6-50bb0c690fe9 | FLAKY | 1/3 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f539-0f34-4233-b49f-a1d8f5323962 | FLAKY | 1/3 |
| rptest.tests.datalake.simple_connect_test.RedpandaConnectIcebergTest.test_translating_avro_serialized_records.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f54b-039e-4ea0-858f-4f25d07def12 | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f54b-039c-47ca-bd97-e67be6e3dac2 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61800#0194f539-0f33-48da-a1a6-50bb0c690fe9 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61803">build#61803</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f69c-0782-4a80-b2b3-956b1e980961 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f69c-0780-4965-bfd9-81ca3854161e | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f69c-0781-4818-a649-7df74eb7ce0f | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.NodeWiseRecoveryTest.test_node_wise_recovery.dead_node_count=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f697-41c0-46d0-a978-c550d40c25dd | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f69c-0780-4965-bfd9-81ca3854161e | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61803#0194f69c-0780-4965-bfd9-81ca3854161e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62093">build#62093</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_when_segment_size_smaller_than_chunk_size | ducktape | https://buildkite.com/redpanda/redpanda/builds/62093#01952733-d99c-45a3-a6d9-94da98ad4a69 | FLAKY | 1/2 |
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62093#01952737-3bec-46af-bfd3-6bc6a1680779 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62093#01952733-d99d-421e-a70d-370a764abd1c | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62093#01952733-d99b-4c50-a628-7614ed3a1123 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62314">build#62314</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FLAKY | 1/3 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_compacting_during_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=100.storage_compaction_key_map_memory_kb=131072 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FAIL | 0/20 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FAIL | 0/20 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FAIL | 0/20 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954081-97f3-407f-aac6-635bf27d36ce | FLAKY | 1/3 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954081-97f0-4939-9d74-9a6fd41664e4 | FAIL | 0/20 |
| rptest.transactions.compaction_e2e_test.CompactionE2EIdempotencyTest.test_basic_compaction.initial_cleanup_policy=compact.workload=Workload.TX | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FLAKY | 1/2 |
| rptest.transactions.compaction_e2e_test.CompactionE2EIdempotencyTest.test_basic_compaction.initial_cleanup_policy=delete.workload=Workload.TX_UNIQUE_KEYS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62314#01954084-fd03-4248-9ae6-04bd07df4ace | FLAKY | 1/4 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62339">build#62339</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_creating_and_listing_migrations | ducktape | https://buildkite.com/redpanda/redpanda/builds/62339#01954357-c2f6-444d-8a7c-fa50cca84313 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTestCompactedTopic.test_write.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62339#0195435b-61cc-4d72-84ca-a19dae5272fd | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62339#0195435b-61cc-4d72-84ca-a19dae5272fd | FLAKY | 1/2 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62339#01954357-c2f6-4753-a177-a4c3e016c730 | FLAKY | 1/4 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62339#01954357-c2f4-4f98-a86b-ba51aadf2152 | FLAKY | 1/9 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62381">build#62381</a></summary>


| test_id                                                                                                                                                                | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_with_transactions.cloud_storage_enabled=True                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62381#019545d3-8d64-46f5-ab09-4a472c6d423e | FLAKY         | 1/2      |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62381#019545bf-4a30-4dde-a082-e589ed45c885 | FLAKY         | 1/2      |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62381#019545bf-4a31-4573-a494-b8462c8bcdb4 | FLAKY         | 1/2      |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61480-->
<!--ac784bd6-715e-49f6-9727-f91751bc3eca-->
<!--title-->
## Retry command for [Build#61480](https://buildkite.com/redpanda/redpanda/builds/61480)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_force_reconfiguration_test.py::NodeWiseRecoveryTest.test_node_wise_recovery@{"dead_node_count":2}
```
<!--/body-->

dotnwat commented on src/v/storage/probe.cc:
```diff
@@ -229,6 +229,16 @@ void probe::setup_metrics(const model::ntp& ntp) {
             "corresponds to the number of times the compaction key-offset map "
             "was unable to be built for a single segment."),
           labels),
+        sm::make_counter(
+          "dirty_segment_bytes",
+          [this] { return _dirty_segment_bytes; },
+          sm::description("Number of bytes within dirty segments of the log"),
+          labels),
+        sm::make_counter(
+          "closed_segment_bytes",
+          [this] { return _closed_segment_bytes; },
+          sm::description("Number of bytes within closed segments of the log"),
+          labels),
```
these descriptions sound like the metrics are gauges (values can increase or decrease) and than counters (strictly increase)

WillemKauf replied:
Addressed in base PR https://github.com/redpanda-data/redpanda/pull/24649

dotnwat commented on src/v/cluster/metadata_cache.cc:
```diff
@@ -324,6 +324,11 @@ metadata_cache::get_default_delete_retention_ms() const {
     return config::shard_local_cfg().tombstone_retention_ms();
 }
 
+std::optional<double>
+metadata_cache::get_default_min_cleanable_dirty_ratio() const {
+    return config::shard_local_cfg().min_cleanable_dirty_ratio();
+}
```
i'm wondering if this is something that we want to expose as a topic config right now, or we keep it internal for a while?

WillemKauf replied:
Is there a reason you would like to keep this config internal for the time being?

I could see an argument for releasing all of `min.cleanable.dirty.ratio`/`min.compaction.lag.ms`/`max.compaction.lag.ms` as configs all at once, but I also think there is use for `min.cleanable.dirty.ratio` as a standalone property for now. 

Is this your reasoning as well?

dotnwat replied:
>Is there a reason you would like to keep this config internal for the time being?

just that no one is asking for it

WillemKauf replied:
There have been some internal discussions around the benefits of `min.cleanable.dirty.ratio` reducing read/write amplification, but fair point.

dotnwat replied:
got it, makes sense. presumably you could bail out of the loop early based on some condition. for example, the while loop already does this in the `while` condition. maybe there isn't a natural way to do that, i'm not sure.

WillemKauf replied:
Yeah, we can just check again for the existence of the ntp in the `ntp_by_compaction_heuristic` map, shift the list and continue without compacting so we eventually hit the "front" of the list again. Sub-optimal but least intrusive change.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -260,29 +293,28 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
         _compaction_hash_key_map = std::move(compaction_map);
     }
-    while (!_logs_list.empty()
-           && is_not_set(_logs_list.front().flags, bflags::compacted)) {
```
i don't see us checking/changing the compacted flags now. is that changed?

alternatively, did you consider just sorting _logs_list? you could leave this while loop unchanged, and before it, iterate over the btree, use get() to access the intrusive hook, and then in O(1) move the entries to the head of the list.

WillemKauf replied:
This is a good idea, thanks. Wasn't sure of the best way to manage both of the housekeeping meta as well as the sorted ntps. This sounds like a good solution!

dotnwat replied:
your changes look ok too, but we've had a bunch of concurrency bugs related to that intrusive list in the past. i think it'd be nice to just sort it how we want, and otherwise leave it untouched. wdyt?

WillemKauf replied:
I agree with your proposed changes, I am just wondering how the while loop would remain unchanged since we may not be compacting every log in the `_logs_list` after these changes (i.e, the typical exit condition of `is_not_set()` returning `false` is no longer the desired logic here).

dotnwat replied:
i was under the impression that it woudl be fine to compact any partition in that loop, we were only interested in which order compaction occurred

WillemKauf replied:
With this PR, the changes (as they stand) do two things:

1. The housekeeping scan no longer indiscriminately compacts every partitions, but instead only compacts partitions with a `dirty_ratio` above `min.cleanable.dirty.ratio` (if it is set at the cluster or topic level).
2. The ordering of compaction for partitions is decided using the `dirty_ratio` as a heuristic.

So, we aren't looking to compact every partition in that loop.

WillemKauf replied:
Sorry, it looks like my original comment got deleted accidentally, for context I wrote:

> I agree with your proposed changes, I'm just wondering how the while loop would remain unchanged when we no longer are looking to compact every log in `_logs_list` (i.e, the logic of `is_not_set()` for the front entry of `_logs_list` returning `false` no longer applies.)

dotnwat replied:
yeh i think we could bail out of the loop (or skip items)?

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -973,6 +973,19 @@ configuration::configuration()
+  , min_cleanable_dirty_ratio(
+      *this,
+      "min_cleanable_dirty_ratio",
+      "The minimum ratio between between the number of bytes in \"dirty\" "
```
```suggestion
      "The minimum ratio between the number of bytes in \"dirty\" "
```

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -973,6 +973,19 @@ configuration::configuration()
+      "min_cleanable_dirty_ratio",
+      "The minimum ratio between between the number of bytes in \"dirty\" "
+      "segments and the total number of bytes in closed segments that must be "
+      "reached before a partition's log is considered eligible for compaction, "
```
```suggestion
      "reached before a partition's log is eligible for compaction "
```

asimms41 commented on tests/rptest/tests/describe_topics_test.py:
```diff
@@ -313,6 +313,17 @@ def test_describe_topics_with_documentation_and_types(self):
+                config_type="DOUBLE",
+                value="-1",
+                doc_string=
+                "The minimum ratio between between the number of bytes in \"dirty\" "
```
```suggestion
                "The minimum ratio between the number of bytes in \"dirty\" "
```

asimms41 commented on tests/rptest/tests/describe_topics_test.py:
```diff
@@ -313,6 +313,17 @@ def test_describe_topics_with_documentation_and_types(self):
+                doc_string=
+                "The minimum ratio between between the number of bytes in \"dirty\" "
+                "segments and the total number of bytes in closed segments that must be "
+                "reached before a partition's log is considered eligible for compaction, "
```
```suggestion
                "reached before a partition's log is eligible for compaction, "
```
```suggestion
                "reached before a partition's log is considered eligible for compaction "
```

WillemKauf commented:
Force push to:
* rebase to `upstream/dev` [post base PR merge ](https://github.com/redpanda-data/redpanda/pull/24649)

WillemKauf committed:
`storage`: add compaction functions to `log_manager_accessor`

WillemKauf committed:
`storage`: add `compaction_scheduling` fixture test

WillemKauf commented:
Force push to:
* Use map of `ntp_to_compaction_heuristic` to sort `_logs_list` in `log_manager.cc` for compaction ordering
* Add `compaction_scheduling` storage fixture test

WillemKauf commented on src/v/storage/log_manager.cc:
```diff
@@ -270,6 +326,11 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
 
         _logs_list.shift_forward();
 
+        if (!ntp_to_compaction_heuristic.contains(
```
I am thinking :thinking: that we could now (because of these changes) have a scheduling race which could lead to an infinite loop in the `while` loop here:

0. `_logs_list` contains some arbitrary number of partitions `N`.
1. Only one partition (`A`) is determined as compactible thanks to compaction heuristic and is added to `ntp_to_compaction_heuristic` map above (Therefore, we don't hit the early return due to `ntp_to_compaction_heuristic.empty()`)
3. During a scheduling point, perhaps during `apply_segment_ms()`, partition `A` is removed from `_logs_list`.
4. We now have an infinite loop in the compacting `while` loop, since the only partition `A` meant for compacting has been removed and `is_not_set(_logs_list.front().flags, bflags::compacted)` will never evaluate to `false`.

Perhaps we should add a new `log_housekeeping_meta::bitflag::compaction_checked`, which is set for every partition whether it is compacted or not, and check that in the `while` loop condition instead.

(Or perhaps that logic is what the current `log_housekeeping_meta::bitflag::compacted` flag should cover, and this `continue` statement just needs to be moved below L334. As a counterpoint to this, there may be a good reason to have a bitflag for both `compacted` and `compaction_checked`.)

WillemKauf replied:
I.e,

```cpp
    while (
      !_logs_list.empty()
      && is_not_set(_logs_list.front().flags, bflags::compaction_checked)) {
        if (_abort_source.abort_requested()) {
            co_return;
        }

        auto& current_log = _logs_list.front();

        _logs_list.shift_forward();

        current_log.flags |= bflags::compaction_checked;

        if (!ntp_to_compaction_heuristic.contains(
              current_log.handle->config().ntp())) {
            continue;
        }

        current_log.flags |= bflags::compacted;
```

WillemKauf replied:
Added `compaction_checked` flag and now the code looks like the above snippet.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61769-->
<!--57fe059a-1c0c-4bd2-a27f-8fcd637c59e7-->
<!--title-->
## Retry command for [Build#61769](https://buildkite.com/redpanda/redpanda/builds/61769)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/mirror_maker_test.py::TestMirrorMakerService.test_consumer_group_mirroring@{"source_type":"kafka"}
```
<!--/body-->

WillemKauf committed:
`storage`: add `compaction_checked` flag to `bitflags`

The housekeeping loop will no longer compact every partition indiscriminately,
but instead evaluate which partitions are worth compacting via a heuristic.

Add a new `bitflag` to `log_housekeeping_meta::bitflags` that indicates
the meta has been evaluated for compaction (but does not indicate that it
has been compacted.)

WillemKauf committed:
`storage`: use `compaction_checked` flag in compacting loop

WillemKauf commented:
Force push to:
* Make recommended docs changes
* Add `log_housekeeping_meta::bitflags::compaction_checked` and use it in main compacting loop in `log_manager.cc`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61775-->
<!--f93204fc-7161-4060-8215-1adbdc72136e-->
<!--title-->
## Retry command for [Build#61775](https://buildkite.com/redpanda/redpanda/builds/61775)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/mirror_maker_test.py::TestMirrorMakerService.test_consumer_group_mirroring@{"source_type":"kafka"}
```
<!--/body-->

WillemKauf commented:
`test_compat` is failing, and it's a little strange. Every so often we get a somewhat truncated `double` back from `serde`:

Expected: `min_cleanable_dirty_ratio: {0.9553594929310676}`

Decoded: `min_cleanable_dirty_ratio: {0.9553594929310675}`

This is the only `diff` between `expected` and `decoded` in the `serde` test.

Do we have a `serde` bug with double floating-point precision values?

WillemKauf committed:
`kafka`: use `conditional_t` for floating point tristate

These call sites were previously hardcoded to `int64_t` for the
type used in a `boost::lexical_cast`.

With the advent of the first `tristate<double>`, this is no longer
a valid default.

Use a conditional type in order to account for the possibility of
a floating point tristate value conversion.

WillemKauf committed:
`rptest`: add `min_cleanable_dirty_ratio` alter config test

With some minor modifications in `kafka_cli_tools.py` and `types.py`.

WillemKauf commented:
Force push to:
* Add `min_cleanable_dirty_ratio_validator` to `kafka` config utils
* Add `test_min_cleanable_dirty_ratio_validation` ducktape test
* Represent `min_cleanable_dirty_ratio` cluster property as a decimal, not a percentage to [align with Kafka](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#min-cleanable-dirty-ratio)

WillemKauf commented on src/v/storage/log_manager.cc:
```diff
@@ -274,6 +330,12 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+
+        if (!ntp_to_compaction_heuristic.contains(
+              current_log.handle->config().ntp())) {
+            continue;
```
this could probably be `co_return;` instead

dotnwat replied:
why? are you implying that once the continue triggers, all the remainder will also continue?

WillemKauf replied:
Yes, because all logs that should be compacted will have been pushed to the front of the `_logs_list`.

WillemKauf commented:
Converting to draft PR. This should not be merged until issues with book-keeping in PR https://github.com/redpanda-data/redpanda/pull/25076 are fixed.

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflicts

WillemKauf commented on src/v/config/configuration.cc:
```diff
@@ -993,6 +993,19 @@ configuration::configuration()
+      {.needs_restart = needs_restart::no,
+       .example = "0.5",
+       .visibility = visibility::user},
+      std::nullopt,
```
This default should be changed to be `0.5`, not disabled. https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#min-cleanable-dirty-ratio

WillemKauf replied:
(Also should set this default for internal topics)

dotnwat replied:
1. presumably there is also a time component: reaches dirty ratio, or some time limit has expired?
2. this is nullopt, did you change it to 0.5 in some later commit?

WillemKauf replied:
>presumably there is also a time component: reaches dirty ratio, or some time limit has expired?

You are referring to the [`max.compaction.lag.ms`](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#max-compaction-lag-ms), which will need to be a follow up work.

> this is nullopt, did you change it to 0.5 in some later commit?

I haven't pushed an update to this yet.

dotnwat replied:
> You are referring to the [max.compaction.lag.ms](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#max-compaction-lag-ms), which will need to be a follow up work.

cool, we have a ticket for this?

WillemKauf replied:
>cool, we have a ticket for this?

https://redpandadata.atlassian.net/browse/CORE-8994

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflicts

WillemKauf commented on src/v/cluster/types.h:
```diff
@@ -568,7 +568,7 @@ struct property_update<tristate<T>>
   : serde::envelope<
       incremental_topic_updates,
-      serde::version<8>,
+      serde::version<9>,
```
might be able to leave this as `serde::version<8>`, need to check tagged version history

dotnwat commented on src/v/kafka/server/handlers/configs/config_utils.h:
```diff
@@ -795,7 +796,9 @@ void parse_and_set_tristate(
     }
     // set property value
     if (op == config_resource_operation::set) {
-        auto parsed = boost::lexical_cast<int64_t>(*value);
```
kinda curious why this wasn't just `lexical_cast<T>(*value)` from the beginning?

WillemKauf replied:
Because `T` can be something like `std::chrono::milliseconds` for time-based properties, and `lexical_cast<>` isn't defined for those types.

We take the parsed value from `lexical_cast<>` and produce the property type from it afterwards in this function `property.value = tristate<T>(std::make_optional<T>(parsed));`.

dotnwat replied:
gotcha, makes sense

dotnwat commented on src/v/kafka/server/tests/alter_config_test.cc:
```diff
@@ -749,6 +749,7 @@ FIXTURE_TEST(
       "redpanda.iceberg.mode",
       "redpanda.leaders.preference",
       "delete.retention.ms",
+      "min.cleanable.dirty.ratio",
```
do we have special procedures for adding new topic configs? i'm thinking specifically about cloud: do we only allow a specific set to be configurable?

WillemKauf replied:
>do we have special procedures for adding new topic configs?

Not that I've seen, just the typical plumbing through the `kafka`/`config`/`storage` layers.

>do we only allow a specific set to be configurable?

Good question, I don't think this set is defined in the Core though? Might need to fact check myself on that.

dotnwat replied:
>  I don't think this set is defined in the Core though

right, i mean, does core have a procedure that invovles coordinating with cloud for example

WillemKauf replied:
>right, i mean, does core have a procedure that invovles coordinating with cloud for example

I'll reach out

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -260,8 +262,9 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
-           && is_not_set(_logs_list.front().flags, bflags::compacted)) {
+    while (
+      !_logs_list.empty()
+      && is_not_set(_logs_list.front().flags, bflags::compaction_checked)) {
```
i don't get it. what does `bflags::compacted` mean now and why are we changing the flags used here? commit message doesn't have any explanation, and it doesn't seem obvious to me what's happening.

WillemKauf replied:
I felt that we should leave `bflags::compacted` to indicate that a log's handle WAS compacted for use cases in the future, and so I added a new flag `bflags::compaction_checked` to indicate that the handle in the log list was considered for compaction.

Otherwise, we get into a situation where the original exit condition doesn't hold. I explained it better in [a comment here](https://github.com/redpanda-data/redpanda/pull/24991#discussion_r1948296081), but I could amend the commit message with more detail if you feel it is lacking.

dotnwat replied:
i see. it makes more sense after reading later commits where the compaction_checked and compaction flags are not always set to together like they are in this commit which introduces the flag.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -250,8 +253,61 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+        co_return;
+    }
+
+    auto sort_by_heuristic = [&ntp_to_compaction_heuristic](
```
just spitballing, but would it be a bit simpler to do something like this:

```
btree_map<double, chunked_vector<log_housekeeping_meta*>> index;
for_each(meta in log_list) {
  index.insert(meta->handle->dirty, &meta);
}
...
for each value in index (index is ordered traversal by dirty ratio)
  for each meta in value
    meta.link.erase();
    logs_list.push_back(meta)
```

or something like that?

WillemKauf replied:
Hm. How about

```cpp
    using compaction_heuristic_t = double;
    btree_map<
      compaction_heuristic_t,
      chunked_vector<log_housekeeping_meta*>>
      index;

    while (!_logs_list.empty()
           && is_not_set(_logs_list.front().flags, bflags::lifetime_checked)) {
        auto& current_log = _logs_list.front();
        _logs_list.shift_forward();

        if (should_compact_log(current_log.handle)) {
            // Order ntps by compaction heuristic.
            // Currently, this is just the dirty ratio.
            auto compute_compaction_heuristic =
              [](ss::shared_ptr<log> l) -> compaction_heuristic_t {
                return l->dirty_ratio();
            };

            auto compaction_heuristic_weight = compute_compaction_heuristic(
              current_log.handle);
            index[compaction_heuristic_weight].push_back(&current_log);
        }
    }

    for (const auto& [weight, vec] : index) {
        for (auto meta_ptr : vec) {
            meta_ptr->link.unlink();
            _logs_list.push_front(*meta_ptr);
        }
    }
```

That way, we're only iterating & sorting over the `meta`s that are going to be compacted.

(I think `unlink()` is the `intrusive_link` function we want)

WillemKauf replied:
I think we were also concerned about concurrency with `_logs_list`/logs being removed, which is why I didn't store a reference/pointer to a `log_housekeeping_meta` object originally (opting for a similar approach in the urgent gc, where the `ntp` of the log is stored instead)

dotnwat replied:
> absl::flat_hash_map<

i assume you meant to use an ordered data structure? and it might also make sense to round to nearest integer.

dotnwat replied:
> I think we were also concerned about concurrency with _logs_list/logs being removed, which is why I didn't store a reference/pointer to a log_housekeeping_meta object originally (opting for a similar approach in the urgent gc, where the ntp of the log is stored instead)

yes, definitely. but in this case we are performing all of the mutations without any scheduling points, right?

WillemKauf replied:
> i assume you meant to use an ordered data structure? and it might also make sense to round to nearest integer.

Oops yeah, corrected. But it doesn't change the rest of my example.

dotnwat replied:
>     _logs_list.sort(sort_by_heuristic);

like i would assume that `sort` is just a loop of unlink/inserts etc...

WillemKauf replied:
> yes, definitely. but in this case we are performing all of the mutations without any scheduling points, right?

Currently we are building the heuristic index in the same loop as where we `co_await current_log.handle->apply_segment_ms();`, but we could also separate it out into two loops. we can also move the compaction map initialization code above this loop to avoid the (albeit rare) scheduling point of `co_await compaction_map->initialize(compaction_mem_bytes);`

dotnwat replied:
good point about apply_segment_ms i missed that. i'm not opposed to separate loops. but the original formulation is ok too. i think it just depends on how much the simplification is valued. up to you

WillemKauf replied:
I don't really have any strong feelings about the extra loop (which would also necessitate adding a new `bflag::compacted_heuristic_checked`) approach vs the `_logs_list.sort()` approach. I imagine they'd be similarly performant. So maybe I'll leave it for now.

dotnwat replied:
```
    for (auto& log_meta : _logs_list) {
        log_meta.flags &= ~(bflags::compacted | bflags::lifetime_checked);
        log_meta.flags &= ~(
          bflags::compacted | bflags::lifetime_checked
          | bflags::compaction_checked);
    }
```

i guess it could go into this loop which already exists? maybe not?

dotnwat reviewed (commented):
looks good! just a few questions, and one discussion about a possible simplification (but it might be too simple and not realistic).

WillemKauf commented:
Force push to:
* Remove `serde::version<>` bump for both `topic_properties` and `incremental_topic_updates`- these were bumped by a previous PR ahead of tagged release `25.1.0`.
* Change cluster default for `min_cleanable_dirty_ratio` from disabled (`std::nullopt`) to `0.5` to match Kafka.

dotnwat reviewed (dismissed):
lgtm

lets me sure the follow up jira ticket exists (if it doesn't already) for the max collection lag thing

WillemKauf committed:
`storage`: change compaction scheduling in `log_manager`

This approach has the added benefit of not having to sort the entire
`_logs_list`. However, we now book-keep two data structures instead
of one. Pointers to `housekeeping_meta` are held, but there are no concerns
about concurrent removal since no scheduling points exist while the order
is being mutated.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -331,8 +330,7 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
 
-        if (!ntp_to_compaction_heuristic.contains(
-              current_log.handle->config().ntp())) {
+        if (!ntps_to_compact.contains(current_log.handle->config().ntp())) {
```
could you get rid of this lookup data structure by using another flag in the bit set we have in the metadata structure?

WillemKauf replied:
Yep, good idea.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -260,8 +262,64 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+    // This needs to be sorted in ascending order, as we are pushing `log_meta`s
+    // to the front of the `_log_list`.
+    absl::
+      btree_map<compaction_heuristic_t, chunked_vector<log_housekeeping_meta*>>
```
should we truncate the floating point in the index? its like, an uncountable number of keys otherwise hah

WillemKauf replied:
Truncated this to `uint64_t`.

WillemKauf commented:
Force push to:
* add `bflags::should_compact` to schedule compaction.
* Use truncated `compaction_heuristic_t = uint64_t` for `btree_map` structure

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -260,8 +262,65 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+            // Currently, this is just the dirty ratio.
+            auto compute_compaction_heuristic =
+              [](ss::shared_ptr<log> l) -> compaction_heuristic_t {
+                auto res = (100.0 * l->dirty_ratio());
```
nit: does the multiplication affect the outcome?

WillemKauf replied:
> does the multiplication affect the outcome?

Without it, the `static_cast<>` will truncate all values down to 0, so it's a necessity here.

dotnwat replied:
oh ha, of course.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -260,8 +262,65 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+            auto compute_compaction_heuristic =
+              [](ss::shared_ptr<log> l) -> compaction_heuristic_t {
+                auto res = (100.0 * l->dirty_ratio());
+                return static_cast<compaction_heuristic_t>(res);
```
any unintended consequence of rounding up vs down?

WillemKauf replied:
I thought about it and I don't think so. Even if a log's heuristic is rounded down to 0, the `bflags::should_compact` flag is still set for the log meta, so all this rounding really affects is the ordering of partitions for compaction.

dotnwat reviewed (dismissed):
lgtm. i think we need more thought put into the default. the simple answer is to use a setting that retains the existing behavior. using the kafka has the potential to just not apply to redpanda.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62308-->
<!--aeb5d6ab-094b-4697-b435-401c6a8f6a4f-->
<!--title-->
## Retry command for [Build#62308](https://buildkite.com/redpanda/redpanda/builds/62308)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
```
<!--/body-->

WillemKauf commented:
> i think we need more thought put into the default. the simple answer is to use a setting that retains the existing behavior. 

Yeah, that would mean disabled by default. I'm not sure what's better here (maintaining the current status quo, or imposing a new sensible default value).

I think the `min.cleanable.dirty.ratio` is something that all `compact` topics should eventually consider setting, which is why I consider having it not disabled by default to be sensible.

I could maybe see leaving it disabled at the cluster level, but defaulting to create new topics with the kafka value `0.5` as a valid option as well.

WDYT?

dotnwat commented:
> I could maybe see leaving it disabled at the cluster level, but defaulting to create new topics with the kafka value 0.5 as a valid option as well.

maybe choose something "close" to the existing behavior. so if existing behavior is 0, then maybe something like 0.2 or whatever. then we can be sure that the code is executing in tests, but not yet have to tackle any sort of tuning project?

WillemKauf commented:
> maybe choose something "close" to the existing behavior. so if existing behavior is 0, then maybe something like 0.2 or whatever. then we can be sure that the code is executing in tests, but not yet have to tackle any sort of tuning project?

Yeah, 0.2 feels like a good value to start with. Ducktape doesn't actually have any complaints with it set to `0.5` which is cool.

I'm open to whatever you feel is best.

dotnwat commented:
> Yeah, 0.2 feels like a good value to start with. Ducktape doesn't actually have any complaints with it set to 0.5 which is cool.
> I'm open to whatever you feel is best.

cool. let's just leave it at 0.5 and we can discuss further when we add the max lag configuration option, and maybe consider integrating with space management.

WillemKauf commented:
(Force pushed to correct value in `describe_topics_test.py` after change in cluster default from `std::nullopt` to `0.5`)

dotnwat commented:
> (Force pushed to correct value in describe_topics_test.py after change in cluster default from std::nullopt to 0.5)

just curious about this, i don't see any recent force pushes

WillemKauf commented:
It's a bit above our most recent conversation, sorry 🙂

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62314-->
<!--bab97538-1615-4e08-80d5-e75521e587c6-->
<!--85ccd908-4896-4bf8-951e-a2af53c98144-->
<!--title-->
## Retry command for [Build#62314](https://buildkite.com/redpanda/redpanda/builds/62314)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/log_compaction_test.py::LogCompactionTest.compaction_stress_test@{"cleanup_policy":"compact,delete","key_set_cardinality":100,"storage_compaction_key_map_memory_kb":131072}
tests/rptest/tests/log_compaction_test.py::LogCompactionTest.compaction_stress_test@{"cleanup_policy":"compact","key_set_cardinality":1000,"storage_compaction_key_map_memory_kb":10}
tests/rptest/tests/log_compaction_test.py::LogCompactionTest.compaction_stress_test@{"cleanup_policy":"compact,delete","key_set_cardinality":1000,"storage_compaction_key_map_memory_kb":3}
```
<!--/body-->

WillemKauf commented:
Well, maybe ducktape doesn't like `0.5` as the cluster default after all.

I can tweak the failing tests to set their `min_cleanable_dirty_ratio` to `std::nullopt`, if we feel strongly that we want to keep `0.5` as the default moving forward.

WillemKauf committed:
`rptest`: set `min_cleanable_dirty_ratio` for ducktape tests

Most of these tests expect/need unconditional compacting of the log.
Set the `min_cleanable_dirty_ratio` at the topic/cluster level in these
tests to mimic the old compaction scheduling behavior before this
configuration was added.

WillemKauf commented:
Force push to:
* Set `min_cleanable_dirty_ratio` for ducktape tests (these all expect unconditional compacting of the log, so set the property to `0.0` throughout)

dotnwat commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -19,118 +19,42 @@
         self.client().create_topic(self.topic_spec)
 
-    def get_removed_tombstones(self):
-        return self.redpanda.metric_sum(
-            metric_name="vectorized_storage_log_tombstones_removed_total",
-            metrics_endpoint=MetricsEndpoint.METRICS,
-            topic=self.topic_spec.name)
```
why did all of these in this commit seem to just move down 40-50 lines from this position?

WillemKauf replied:
Ugliness of the git diff, the change is simply moving all these helper functions to the base class. Easier to grok if you open up the file sans diff, I think.

dotnwat commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -242,7 +256,6 @@ def compaction_has_completed():
 
         # There should be no dirty segments left
         assert self.get_dirty_segment_bytes() == 0
-        assert self.get_closed_segment_bytes() > 0
```
why is this now removed? i don't see in the base class

WillemKauf replied:
This was actually race-y for the test when parameterized with `compact,delete`. Sometimes the deletion policy would remove all the closed bytes after waiting on compaction, causing this assert to fail. 

I could have done a better job of calling this out here.

dotnwat commented on tests/rptest/clients/kafka_cli_tools.py:
```diff
@@ -158,6 +158,11 @@ def create_topic(self, spec: TopicSpec):
             args += [
                 "--config", f"delete.retention.ms={spec.delete_retention_ms}"
             ]
+        if spec.min_cleanable_dirty_ratio:
```
this won't work as expected for a value of `0`?

WillemKauf replied:
Shoot. Right. Will fix, going to wait on green CI for the existing build first.

dotnwat replied:
can we pull out these last 3 changes and review them separately?

WillemKauf replied:
I pulled out the added test (last two commits)

We still need the changes to existing ducktape tests with setting `min.cleanable.dirty.ratio` to 0.

WillemKauf commented:
Force push to:
* Remove added ducktape test for future PR.

dotnwat commented:
Cool. But what was this change?

https://github.com/redpanda-data/redpanda/compare/851be256a2e6bce3296e5797e83da6b3d2ce0d2c..a9cb0be0ad25b879a00aa69bef64461076c5793b

I don't recall it being mentioned

WillemKauf commented:
That change was an amend to the commit for the new ducktape test which was removed. It's no longer in this PR.

WillemKauf commented:
CI flakes are:

https://redpandadata.atlassian.net/issues/CORE-8500
https://redpandadata.atlassian.net/issues/CORE-9114
https://redpandadata.atlassian.net/issues/CORE-7129

Looks good to me
